Expose maximum_frame_latency

Wumpf opened the pull request:
**Connections**
 * Scratches the surface of https://github.com/gfx-rs/wgpu/issues/2869
 * _Way_ better alternative to setting to `Immediate` as initially proposed here
      * https://github.com/emilk/egui/pull/3714
      * https://github.com/rerun-io/rerun/pull/4575

**Description**
Previously, the swapchain buffer count was hardcoded to be always 3 no matter what is supported. This PR exposes the desired swap chain count as a minimal extension to the wgc presentation setup.

Among other things this allows one now to switch between "Game Mode" and "Classic Game Mode" as classified here
https://www.intel.com/content/www/us/en/developer/articles/code-sample/sample-application-for-direct3d-12-flip-model-swap-chains.html

Prime motivation for this is to reduce perceived lag in egui applications where a buffer count of 2 makes usually more sense (assuming a stable and low workload).

**Testing**
Experimented with buffer count of 2 in egui example on Mac & Windows. Gives a lot less latency.

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `cargo clippy`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
  - [ ] `--target wasm32-unknown-emscripten`
- [ ] Run `cargo xtask test` to run tests.
- [x] Add change to `CHANGELOG.md`. See simple instructions inside file.

Wumpf committed:
Expose `desired_swap_chain_size`

Wumpf committed:
update examples

Wumpf committed:
update changelog

Wumpf committed:
fix missing fixup

emilk commented:
I just want to emphasize what a huge difference changing going from 3 to 2 is on Mac, at least in windowed mode on a 60 Hz monitor. It takes the application from feeling sluggish to feeling responsive (using `AutoVsync`).

When the CPU is fast (e.g. producing a frame in 1ms), `SwapChain::get_next_texture` will block, saving CPU time and preventing tearing.

Interestingly though, when the CPU is too slow (e.g. taking 20ms to prepare a frame), then `SwapChain::get_next_texture` will _not_ block. I'm not sure why, but I like it.

So: I'm planning on making `2` the default for all `eframe` apps

Wumpf committed:
doc fix

emilk commented on CHANGELOG.md:
```diff
@@ -66,6 +66,7 @@ Wgpu now exposes backend feature for the Direct3D 12 (`dx12`) and Metal (`metal`
 - No longer validate surfaces against their allowed extent range on configure. This caused warnings that were almost impossible to avoid. As before, the resulting behavior depends on the compositor. By @wumpf in [#4796](https://github.com/gfx-rs/wgpu/pull/4796)
 - Added support for the float32-filterable feature. By @almarklein in [#4759](https://github.com/gfx-rs/wgpu/pull/4759)
 - wgpu and wgpu-core features are now documented on docs.rs. By @wumpf in [#4886](https://github.com/gfx-rs/wgpu/pull/4886)
+- `SurfaceConfiguration` now exposes `desired_swap_chain_size` which was previously hard-coded to 3. By setting it to 2 you can reduce latency. By picking a large value you can ensure higher throughput. By @emilk & @wumpf in [#????](https://github.com/gfx-rs/wgpu/pull/????)
```
```suggestion
- `SurfaceConfiguration` now exposes `desired_swap_chain_size` which was previously hard-coded to 3. By setting it to 2 you can reduce latency. By picking a large value you can ensure higher throughput. By @emilk & @wumpf in [#4899](https://github.com/gfx-rs/wgpu/pull/4899)
```

emilk committed:
update PR number

Wumpf committed:
formatting

cwfitzgerald reviewed (changes requested):
Marking request review per discussion in call

Wumpf committed:
Change swap_chain_size to maximum_frame_latency
On most platforms this is merely a change in narrative, on DX12 it actually has a practical effect already since we use it to directly set frame latency

Wumpf commented:
code looks mostly now like we want it to be I think! For sure way more expressive of the stated goal and in close alignment with
* #2869

Need to update changelog and test whether `maximum_frame_latency == 1` causes gpu/cpu to run in sync on Vulkan & Metal (hoping to find this answered with 'no, they still run parallel')

emilk commented on wgpu-hal/src/dx12/mod.rs:
```diff
@@ -654,13 +654,18 @@ impl crate::Surface<Api> for Surface {
+        // Nvidia recommends to use 1-2 more buffers than the maximum latency
+        // https://developer.nvidia.com/blog/advanced-api-performance-swap-chains/
+        // For high latency extra buffers seems excessive, so go with a minimum of 3 and beyond that add 1.
+        let swap_chain_buffer = (config.maximum_frame_latency + 1).min(3);
```
won't this make double-buffering impossible?

Wumpf replied:
The number of swap chain buffer isn't _actually_ the thing we care about to begin with: we care about the latency from recording to display and DX12 can set this more directly than using [`SetMaximumFrameLatency`](https://learn.microsoft.com/en-us/windows/win32/api/dxgi/nf-dxgi-idxgidevice1-setmaximumframelatency). Essentially it decouples the "wait on queue is full" from the number of buffers that are used. Which is why @cwfitzgerald actually advocated originally to just always use 3. But naturally, the desired frequency isn't entirely decoupled which is why Nvidia suggests to over provision by at least 1.
To my understanding the major perk of this more principled approach is that instead of blocking on `Present`/buffer flip, y we wait at the begin of a frame using a "waitable object" which is signaled **after** a present is done. See [getframelatencywaitableobject](https://learn.microsoft.com/en-us/windows/win32/api/dxgi1_3/nf-dxgi1_3-idxgiswapchain2-getframelatencywaitableobject). This means that if there's (for whatever reason) a longer wait, you have frame preparation including all the input query much much closer to the actual presenting of the frame.

Timing your frame preparation more accurately and consistently gets you into the even more complex real of "frame pacing" which is all about pacing your frame recording to hit the next screen refresh as closely as possible; there's various mutally inconsistent APIs popping up these days (see #2869) so it's that part of the equation isn't that easy to solve.

Unfortunately, on Vulkan this works only with some extensions. To be implemented. (Oh and since you're wondering now: Rerun doesn't use DX12 right now for a bunch or reasons, we always stick with Vulkan)

I'm still learning this stuff, so maybe @cwfitzgerald can correct me if I got something wrong :)

Wumpf replied:
tl'dr: The trick is that on dx12 `SetMaximumFrameLatency` takes `maximum_frame_latency` as-is.

emilk replied:
…I also misread `.min(3)` as "at minimum 3" 🤦

emilk commented on wgpu-hal/src/dx12/mod.rs:
```diff
@@ -687,7 +692,9 @@ impl crate::Surface<Api> for Surface {
                     },
                     buffer_usage: dxgitype::DXGI_USAGE_RENDER_TARGET_OUTPUT,
-                    buffer_count: config.swap_chain_size,
+                    // Nvidia recommends to use 1-2 more buffers than the maximum latency
+                    // https://developer.nvidia.com/blog/advanced-api-performance-swap-chains/
+                    buffer_count: swap_chain_buffer,
```
duplicate comment

Wumpf replied:
oops

emilk commented on wgpu-hal/src/vulkan/adapter.rs:
```diff
@@ -1838,7 +1838,12 @@ impl crate::Adapter<super::Api> for super::Adapter {
+            // (presumably - we're actually setting the min image count which isn't necessarily the swap chain size)
+            // Instead, we should use extensions when available to wait in present.
+            // See https://github.com/gfx-rs/wgpu/issues/2869
+            maximum_frame_latency: caps.min_image_count.saturating_sub(1)
+                ..=max_image_count.saturating_sub(1),
```
this should also have an `.max(1)` on it, since the docstring for `SurfaceCapabilities::maximum_frame_latency` requires at least `1`

Wumpf replied:
good catch. I checked the spec and I'd now say my comment is actually wrong. `maximum_frame_latency`==0 (if supported) to my understanding now should mean that gpu/cpu are in sequence (although @cwfitzgerald was skeptical about this before). I'll have to experiment though to confirm

emilk commented on wgpu-types/src/lib.rs:
```diff
@@ -5107,13 +5107,16 @@ pub struct SurfaceConfiguration<V> {
+    /// As a consequence either the maximum frame latency is set directly on the swap chain,
+    /// or waits on present to avoid exceeding the maximum frame latency,
+    /// or the swap chain size is set to max-latency + 1.
+    pub desired_maximum_frame_latency: u32,
```
we should make a note on what happens if the user sets this to zero.

Do all backends clamp this to a valid range?

Wumpf replied:
Yeah right now, they all clamp to the range - this is unchanged from the thing we pair programmed on.
I think in the future we want to go as in Connor's original proposal and expose this just like we expose format and size restrictions. We'll then drop the `desired_` part and check again it as with other properties.

Expanding the comment!

daxpedda commented on CHANGELOG.md:
```diff
@@ -66,6 +66,7 @@ Wgpu now exposes backend feature for the Direct3D 12 (`dx12`) and Metal (`metal`
 - No longer validate surfaces against their allowed extent range on configure. This caused warnings that were almost impossible to avoid. As before, the resulting behavior depends on the compositor. By @wumpf in [#4796](https://github.com/gfx-rs/wgpu/pull/4796)
 - Added support for the float32-filterable feature. By @almarklein in [#4759](https://github.com/gfx-rs/wgpu/pull/4759)
 - wgpu and wgpu-core features are now documented on docs.rs. By @wumpf in [#4886](https://github.com/gfx-rs/wgpu/pull/4886)
+- `SurfaceConfiguration` now exposes `desired_swap_chain_size` which was previously hard-coded to 3. By setting it to 2 you can reduce latency. By picking a large value you can ensure higher throughput. By @emilk & @wumpf in [#4899](https://github.com/gfx-rs/wgpu/pull/4899)
```
```suggestion
- `SurfaceConfiguration` now exposes `desired_swap_chain_size` which was previously hard-coded to 3. By setting it to 1 you can reduce latency. By picking a large value you can ensure higher throughput. By @emilk & @wumpf in [#4899](https://github.com/gfx-rs/wgpu/pull/4899)
```
According to the documentation on `SurfaceConfiguration` its now 2 by default and it's recommended to set it to 1 to reduce latency.

Wumpf replied:
yep, as mentiond below I haven't updated the changelog again yet

daxpedda commented on wgpu-types/src/lib.rs:
```diff
@@ -5107,6 +5107,16 @@ pub struct SurfaceConfiguration<V> {
     pub present_mode: PresentMode,
+    /// Desired maximum number of frames that the presentation engine should queue in advance.
+    ///
+    /// Defaults to 2 when created via `wgpu::Surface::get_default_config`.
+    /// Recommended to use 2 (or higher) for high throughput, 1 for low latency.
```
Could we clarify exactly what "high throughput" means?
Maybe it just isn't clear to me because I'm clueless on the topic ...

Wumpf replied:
Hmm I'll try to expand a bit on it, but it's intentionally handwavy :/

Wumpf replied:
rewrote that section now to be a lot more precise :)

Wumpf committed:
Merge remote-tracking branch 'origin/trunk' into expose-swapchain-buffer-count

Wumpf committed:
update changelog, improve comments

Wumpf committed:
fix missing rename, remove unused and incorrectly named variable

Wumpf commented:
Got around to test with Vulkan on windows of when GPU/CPU ending up to working in sync.
My test setup was to render the fractal of wgpu-profiler's demo with more iterations (making it easy to control the gpu load) and do articial cpu load by via thread sleep. I then compared frame times with `maximum_frame_latency` 1 and `maximum_frame_latency` 2.
Turns out with a frame latency setting of 1 the gpu won't have any work between `get_current_texture` and the next queue submit. Or in other words: When I put the sleep _after_  `get_current_texture`, the gpu will be fully idle since there's no work to be done at all. Meaning at the very least need to adjust the documentation.

On DX12 I'm hitting some odd bug making frame time expand to 1 full second, so I don't have test results from that so far.

Not the result I was hoping for. Need to repeat the same on Mac to better understand stuff.

emilk commented:
> On DX12 I'm hitting some odd bug making frame time expand to 1 full second, so I don't have test results from that so far.

Sounds like you might be hitting `const FRAME_TIMEOUT_MS: u32 = 1000;`

Wumpf committed:
iterating on desired_maximum_frame_latency doc

Wumpf committed:
fix macos build

Wumpf commented:
Same behavior on Metal.
Thinking about this some more this starts making more sense to me. Already updated docs accordingly again. I suppose setting the latency to 1 still makes sense for fast rendering ui intense applications, so exposing this is still good. But use of it definitely requires some special care to not completely starve the gpu.

The thing I don't get yet is how one `CAMetalLayer` can do a wait on present. #2869 mentioned one can but I didn't find out how.

Wumpf committed:
iterate on documentation, update changelog

cwfitzgerald reviewed (approved):
Lets party!
